cluster/src/main/java/org/apache/iotdb/cluster/log/CommittedEntryManager.java

```diff
@@ -0,0 +1,128 @@
+    }
+
+    public HardState getHardState() {
+        return hardState;
+    }
+
+    public Snapshot getSnapshot() {
+        return snapshot;
+    }
+
+    public void applySnapshot(Snapshot snap) {
+        long localIndex = snapshot.getLastLogIndex();
+        long snapIndex = snap.getLastLogIndex();
+        if (localIndex >= snapIndex) {
+            logger.info("requested index is older than the existing snapshot");
+            return;
+        }
+        snapshot = snap;
+        entries.clear();
+        Log dummy = new PhysicalPlanLog();
+        dummy.setCurrLogIndex(snap.getLastLogIndex());
+        dummy.setCurrLogTerm(snap.getLastLogTerm());
+        entries.add(dummy);
+    }
+
+    public Long getFirstIndex() {
+        return entries.get(0).getCurrLogIndex() + 1;
+    }
+
+    public Long getLastIndex() {
+        return entries.get(0).getCurrLogIndex() + entries.size() - 1;
+    }
+
+    public long getTerm(long index) {
+        long offset = entries.get(0).getCurrLogIndex();
+        if (index < offset) {
+            return -1;
+        }
+        if ((int) (index - offset) >= entries.size()) {
+            return -1;
+        }
+        return entries.get((int) (index - offset)).getCurrLogTerm();
+    }
+
+    public List<Log> getEntries(long low, long high) throws EntryCompactedException, EntryUnavailableException {
+        long offset = entries.get(0).getCurrLogIndex();
+        if (low <= offset || entries.size() == 1) {
+            throw new EntryCompactedException();
+        }
+        if (high > getLastIndex() + 1) {
+            logger.error("entries high ({}) is out of bound lastIndex ({})", high, getLastIndex());
+            throw new EntryUnavailableException();
+        }
+        return entries.subList((int) (low - offset), (int) (high - offset));
+    }
+
+    public void compactEntries(long compactIndex) throws EntryCompactedException, EntryUnavailableException {
+        long offset = entries.get(0).getCurrLogIndex();
+        if (compactIndex < offset) {
+            throw new EntryCompactedException();
+        }
+        if (compactIndex > getLastIndex()) {
+            logger.error("compact ({}) is out of bound lastIndex ({})", compactIndex, getLastIndex());
+            throw new EntryUnavailableException();
+        }
+        int index = (int) (compactIndex - offset);
+        PhysicalPlanLog dummy = new PhysicalPlanLog();
+        dummy.setCurrLogIndex(entries.get(index).getCurrLogTerm());
+        dummy.setCurrLogTerm(entries.get(index).getCurrLogIndex());
+        entries.clear();
+        entries.add(dummy);
+    }
+
+    public void append(List<Log> entries) {
+        if (entries.size() == 0) {
+            return;
+        }
+        long first = getFirstIndex();
+        long last = entries.get(0).getCurrLogIndex() + entries.size() - 1;
+        if (last < first) {
+            return;
+        }
+        if (first > entries.get(0).getCurrLogIndex()) {
```

Comment:
I have added a function `getDummyIndex()` for `this.entries.get(0).getCurrLogIndex()`.
BTW, the `entries.get(0).getCurrLogIndex()`  here  represents the parameter `entries`, not the local `entries`. To avoid ambiguity, I have renamed parameter `entries` to `appendingEntries`.
Thanks!